### ISSUE TITLE
Any precondition for the provided 2d points and 3d points

### ISSUE THREAD

**Issue by hetolin** [dylan-campbell/bpnpnet#3]
Hi, thank you for sharing your code.  I wonder that if the 2d points are sampled from object A in the image I, while the 3d points are sampled from object B in the image I. Could the bpnpnet be suitable for such an experimental setting? Since it seems that there is no correspondence between the 2d points and 3d points in such conditions.

Or is there any requirements/precondition for the provided 2d points and 3d points? (such as the number of the provided 2d and 3d points at least, the sample range of the 2d/3d points, etc.)

**Comment by dylan-campbell** [comment]
Hi @hetolin, no, it couldn't be used in such a setting. Some correspondences need to exist, even though they aren't known to begin with. That being said, the model will still try to align object A and B, which might be useful in some contexts (though not for camera pose estimation).

Yes, there are some requirements on the number and layout of 2D and 3D points, same as for any PnP algorithm. At least 4 2D-3D pairs are required at minimum, and they can't be degenerate.

**Comment by hetolin** [comment]
Thank you for your prompt reply. As you mentioned above, the network takes the paired 2D-3D points as input. The video provided in your paper mentioned that the lidar data(3d points) and the image data(2d points) are not paired, so the bpnpnet could not be used in such a setting?  

That is to say, the number of 2d points and 3d points is required to be equal?

**Comment by dylan-campbell** [comment]
No, the number of 2D and 3D points aren't required to be equal. And yes, lidar + image data is suitable. The requirement is just that there is some overlap between the 2D points and the 3D points: some of the 2D-3D pairs must correspond approximately. So if you have taken a lidar scan of an object and image points of the same object, that is satisfactory.

**Comment by hetolin** [comment]
Thank you, I see what you mean. But there is a last one question left: As for an object M, I get the surface 2d pixel points and the 3d points sampled from the whole 3D model. And if I sent the partial 2d points and full 3d points into the bpnpnet.  That is, 2d points are a subset of the 3d points,  and the number of 2d points is much less than that of 3d points(supposed that the number of 2d points is half of that of 3d points). In such a setting, would it degrade the performance or influence the accuracy of camera pose estimation? 
Or does it is necessary to add additional data to finetune the network if the test scene is far different from the training sets?

**Comment by dylan-campbell** [comment]
Hi @hetolin, yes the performance degrades as the overlap decreases. You can get an idea of the performance degradation in the supplementary material (Fig 1 for random outliers, Fig 2 for structured outliers - which is what you are talking about). These results show robustness to partial overlap despite only training with 1:1 data, so fine-tuning the model or retraining it with partially-overlapping data would probably improve the robustness even further.

**Comment by hetolin** [comment]
I got it! Thank you very much for your detailed reply.
